Add HTTP/3 support (#1629)
microsoft/reverse-proxy: docs/docfx/articles/http3.md

# HTTP/3

## Introduction

> To enable HTTP/3 protocol on YARP you should set up HttpClient and Kestrel.

Reviewer: ```suggestion
YARP supports HTTP/3 for inbound and outbound connections using the HTTP/3 preview support in .NET 6. To enable the HTTP/3 protocol in YARP you need to:
- Configure inbound connections in Kestrel
- Configure outbound connections in HttpClient 
- Enable preview features
```

## Set up HTTP/3 on Kestrel

Protocols are required in the listener options:
```C#
var myHostBuilder = Host.CreateDefaultBuilder(args);
myHostBuilder.ConfigureWebHostDefaults(webHostBuilder =>
    {
        webHostBuilder.ConfigureKestrel(kestrel =>
        {
            kestrel.ListenAnyIP(443, portOptions =>
            {
#if NET6_0_OR_GREATER
                portOptions.Protocols = HttpProtocols.Http1AndHttp2AndHttp3;
#endif
            portOptions.UseHttps();
        });
    });
    webHostBuilder.UseStartup<Startup>();
});
```

> For correct project build with `HttpProtocols.Http1AndHttp2AndHttp3` preview features need to be enabled:

Reviewer: ```suggestion
To use HTTP/3 with .NET 6, preview features need to be enabled via a setting in the project file:
```

```proj
<PropertyGroup>
  <EnablePreviewFeatures>True</EnablePreviewFeatures>
</PropertyGroup>
```

## HttpClient

There is the switch which enables HTTP/3 either programmatically or in project features:

> ```C#

Reviewer: Rework to show 2 snippets, one for project, the other in code.

# HTTP/3

## Introduction
YARP 1.1 supports HTTP/3 for inbound and outbound connections using the HTTP/3 preview support in .NET 6. To enable the HTTP/3 protocol in YARP you need to:
- Configure inbound connections in Kestrel
- Configure outbound connections in HttpClient 
- Enable preview features

## Set up HTTP/3 on Kestrel

Protocols are required in the listener options:
```C#
var myHostBuilder = Host.CreateDefaultBuilder(args);
myHostBuilder.ConfigureWebHostDefaults(webHostBuilder =>
    {
        webHostBuilder.ConfigureKestrel(kestrel =>
        {
            kestrel.ListenAnyIP(443, portOptions =>
            {
                portOptions.Protocols = HttpProtocols.Http1AndHttp2AndHttp3;
                portOptions.UseHttps();
            });
        });
    webHostBuilder.UseStartup<Startup>();
    });
```
To use HTTP/3 with .NET 6, preview features need to be enabled via a setting in the project file:

> ```proj

Reviewer 2: ```suggestion
```csproj
```

<PropertyGroup>
  <EnablePreviewFeatures>True</EnablePreviewFeatures>
</PropertyGroup>
```

## HttpClient

There is the switch which enables HTTP/3 either programmatically:
```C#
AppContext.SetSwitch("System.Net.SocketsHttpHandler.Http3Support", true);
```

> or in project feature:

Reviewer 2: ```suggestion
or in project file:
```

```csproj
<RuntimeHostConfigurationOption Include="System.Net.SocketsHttpHandler.Http3Support" Value="true" />
```

> In addition to this default version of HttpRequest should be replaced by "3", find more details about [HttpRequest configuration](http-client-config.md#httprequest).

Reviewer 2: ```suggestion
In addition to this, the default version of HttpRequest should be replaced by "3", find more details about [HttpRequest configuration](http-client-config.md#httprequest).
```
